### ISSUE TITLE
tf.image.non_max_suppression bug when iou_threshold=0.0

### ISSUE THREAD

**Issue by ziky90** [tensorflow/tensorflow#34268]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): both Mac and Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.15.0
- Python version: 3.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
I'm running following code:
```
boxes = [[ 90., 480., 106., 496.],
         [269.33333333, 536.33333333, 282.66666667, 549.66666667],
         [382.33333333, 556.33333333, 395.66666667, 569.66666667]]
result = tf.image.non_max_suppression(boxes, [1, 1, 1], 3, iou_threshold=0.0)
    with tf.Session() as sess:
        out = sess.run([result])
        indexes = out[0]
print(len(indexes))
```
Which prints out: 1 in `tensorflow==1.15.0`.

**Describe the expected behavior**
Expected behaviour is printing 3 as in tensorflow==1.13.1.
I've found a workaround for tensorflow==1.15.0 as setting io_threshold to some very small number, eg. `iou_threshold=0.001`, but the behaviour, probably a bug when `iou_threshold=0.0` is very very unexpected.

**Code to reproduce the issue**
```
boxes = [[ 90., 480., 106., 496.],
         [269.33333333, 536.33333333, 282.66666667, 549.66666667],
         [382.33333333, 556.33333333, 395.66666667, 569.66666667]]
result = tf.image.non_max_suppression(boxes, [1, 1, 1], 3, iou_threshold=0.0)
    with tf.Session() as sess:
        out = sess.run([result])
        indexes = out[0]
print(len(indexes))
```

This seems to me probably as a bug in case when iou_threshold=0.0. The bug was not present at tensorflow==1.13.1, but it's in tensorflow==1.15.0.

**Comment by ziky90** [comment]
I see that actually this "corner case" is probably not tested in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_test.py it looks to me that the functionality is only tested for `iou_threshold=0.5`.

**Comment by ravikyram** [comment]
I tried in colab and was able to reproduce the issue. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/79a3a8395a6edb27e870bac86cd9035f/untitled369.ipynb). Thanks!

**Comment by ziky90** [comment]
@ravikyram btw. the very same bug seems to be also in tensorflow==2.0.0, so maybe you can add also label `TF 2.0`?

You can reproduce it by:
```
boxes = [[ 90., 480., 106., 496.],
         [269.33333333, 536.33333333, 282.66666667, 549.66666667],
         [382.33333333, 556.33333333, 395.66666667, 569.66666667]]
indexes = tf.image.non_max_suppression(boxes, [1, 1, 1], 3, iou_threshold=0.0)
print(len(indexes))
```
which prints 1

and
```
boxes = [[ 90., 480., 106., 496.],
         [269.33333333, 536.33333333, 282.66666667, 549.66666667],
         [382.33333333, 556.33333333, 395.66666667, 569.66666667]]
indexes = tf.image.non_max_suppression(boxes, [1, 1, 1], 3, iou_threshold=0.001)
print(len(indexes))
```
which prints 3

In both cases expected result is 3

**Comment by rmothukuru** [comment]
Could reproduce the issue with 2.0 as well. Here is the [Gist](https://colab.sandbox.google.com/gist/rmothukuru/6e27abeb2ef09c6384fa64e038371a47/untitled369.ipynb#scrollTo=sPsaV2E7ua0S).

**Comment by alextp** [comment]
@jch1 I think you wrote this code, can you take a look?

**Comment by ziky90** [comment]
@alextp @jch1 any progress looking into this issue?

**Comment by gadagashwini** [comment]
Was able to reproduce the issue with Tf-nightly==.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/1de3e0bedb94476be842fcc8923f2e8a/untitled470.ipynb). Thanks

**Comment by alextp** [comment]
@danielcdh can you take over this issue? (you should be in the TF github org)

**Comment by danielcdh** [comment]
@jinliangwei will work on this issue. Both of us are in the process to be added to TF github org.

**Comment by jinliangwei** [comment]
The real question here is what is the correct definition for "iou_threshold"?

According to Bodla et al. (http://www.cs.umd.edu/~bharat/snms.pdf), NMS suppresses box_a or box_b depending on their score, if iou(box_a, box_b) >= iou_threshold, which is the current implementation does, i.e., select only one box when iou_threshold = 0.0 in the above example.

In previous implementations, NMS suppress box_a or box_b, if iou(box_a, box_b) > iou_threshold, and thus select three boxes when iou_threshold = 0.0 in the above example.

What definition do we really want? @jch1 Would you please comment?

**Comment by amahendrakar** [comment]
Issue persists in [TF v2.2](https://colab.research.google.com/gist/amahendrakar/d5e885eb1cef9d2a157c5ca9c9e3c004/34268.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/22694d3212c28afe0a8f2208bf0b8c4b/34268-tf-nightly.ipynb) i.e. v2.3.0-dev20200527. Please find the attached gist. Thanks!

**Comment by bhack** [comment]
It is fixed in nightly.

**Comment by rmothukuru** [comment]
@ziky90,
As per [bhack's comment](https://github.com/tensorflow/tensorflow/issues/34268#issuecomment-840717628) this bug is fixed in **`Nightly Version of Tensorflow (2.6.0-dev20210516)`**. Please find [the Gist](https://colab.research.google.com/gist/rmothukuru/60ff1d38ba9af0d7ffffbd452f07cbdf/34268-tf-nightly.ipynb#scrollTo=JrpVby8YUwdF) of the working code. 

Can you please let us know if we can close this issue? Thanks!
